src/main/java/com/hierynomus/smbj/session/Session.java

```diff
@@ -72,6 +72,37 @@ public Session(Connection connection, AuthenticationContext auth, SMBEventBus bu
     public void init(SMB2SessionSetup setup) {
         this.guest = setup.getSessionFlags().contains(SMB2SessionSetup.SMB2SessionFlags.SMB2_SESSION_FLAG_IS_GUEST);
         this.anonymous = setup.getSessionFlags().contains(SMB2SessionSetup.SMB2SessionFlags.SMB2_SESSION_FLAG_IS_NULL);
+        validateAndSetSigning(setup);
+        if (guest || anonymous) {
+            packetSignatory.init(null); // De-initialize the signatory if it was initialized with a bogus key during the auth
+        }
+    }
+
+    /**
+     * [MS-SMB2] 3.2.5.3.1 Handling a New Authentication
+     * @param setup
+     */
+    private void validateAndSetSigning(SMB2SessionSetup setup) {
+        boolean requireMessageSigning = connection.getConfig().isSigningRequired();
+        boolean connectionSigningRequired = connection.getConnectionInfo().isServerRequiresSigning();
+
+        if (requireMessageSigning || connectionSigningRequired) {
+            signingRequired = true;
+        }
+
+        if (anonymous) {
+            // If the security subsystem indicates that the session was established by an anonymous user, Session.SigningRequired MUST be set to FALSE.
+            signingRequired = false;
+        }
+        if (guest && connection.getConfig().isSigningRequired()) {
+            throw new IllegalStateException("Cannot require message signing when authenticating with a guest account");
```

Comment:
Could we use a specific `SmbRuntimeException` for this?

Reply:
Yes, will do